### PR TITLE
Plugin to search for commit messages in your _local_ GIT repositories

### PR DESCRIPTION
**Pull request by ewasser**
I've created a plugin to search for the commit messages in your _local_ GIT repositories.

It will look for repositories in `~/git/*/*` and `~/src/*/*`. The first search parameter is the name of the repository in your directory and the second search parameter will be used to grep through all the `git log` from the choosen repository.

# EXAMPLE #

`git foo` - List all our local git with `foo` in the path

After using the `TAB` you can search in the choosen directory for a commit message.

`git foo align` - List all commits with the word `align` in it.

The commit message will be copied to the clipboard.

### CODE REVIEW

**Comment by ManuelSchneid3r** [comment]
Hmm this is pretty ridgid. I dont think that a lot of ppl have these dir. I have ~/git though :smile:

**Comment by ewasser** [comment]
No problem, I've made it configurable during a little file `~/.config/albert/org.albert.extension.gitlog/config.yaml`. This is a little YAML-file with your git directories in it, here's little example:

    ---
    git-directories:
    - ~/git/*/*
    - ~/src/*/*

File: GitLog.py [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+class GitConfig:
+    def __init__(self, **kwargs):
+        super().__init__()
+        self.config = configparser.ConfigParser()
+
+    def read(self, directory):
+        configuration_file = os.path.join(directory, '.git/config')
+        self.config.read(configuration_file)
+
+### SHAMELESS CODE REUSE FROM https://github.com/johnkchiu/GitLogParser/blob/master/gitLogParser.py
```
Comment (idkCpp) [inline GitLog.py]: This Repository has no licence attached. I just wanted to bring it up.

File: GitLog.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+                if len(commit) != 0:		## new commit, so re-initialize
+                    self.commits.append(commit)
+                    commit = {}
+                commit = {'hash' : re.match('commit (.*)', nextLine, re.IGNORECASE).group(1) }
+            elif bool(re.match('merge:', nextLine, re.IGNORECASE)):
+                # Merge: xxxx xxxx
+                pass
+            elif bool(re.match('author:', nextLine, re.IGNORECASE)):
+                # Author: xxxx <xxxx@xxxx.com>
+                m = re.compile('Author: (.*) <(.*)>').match(nextLine)
```
Comment (idkCpp) [inline GitLog.py]: Why do you compile the regexp here and don't compile it the lines before or after?
Skip the compiling. According to https://stackoverflow.com/a/452143 it Regexes are cached internally.

File: GitLog.py [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+                commit['author'] = m.group(1)
+                commit['email'] = m.group(2)
+            elif bool(re.match('date:', nextLine, re.IGNORECASE)):
+                # Date: xxx
+                m = re.compile('Date: +(.*)').match(nextLine)
```
Comment (idkCpp) [inline GitLog.py]: Skip compiling

File: GitLog.py [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+               source = source[len(prefix):]
+
+           yield source, target
+
+import configparser
+import os.path
+
+class GitConfig:
+    def __init__(self, **kwargs):
+        super().__init__()
```
Comment (idkCpp) [inline GitLog.py]: What does this do? This class does not have a Base Class.

**Comment by ManuelSchneid3r** [comment]
There is a new [community repository for python extensions](https://github.com/albertlauncher/python_community) now. Please send the PR there. Sorry for the inconvenience. I am just trying to reduce the workload here and make you happy with contemporary merges. If you want to join the party (write access) ask for it in the community chats.

**Pull request closed without merging**